src/transport/tests/TestGroupMessageCounter.cpp

```diff
@@ -0,0 +1,330 @@
+            NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+        }
+        fabricIndex++;
+    }
+
+    // Try re-adding the previous peer without any error
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+}
+
+void PeerRetrievalTest(nlTestSuite * inSuite, void * inContext)
+{
+    NodeId peerNodeId                              = 1234;
+    FabricIndex fabricIndex                              = 1;
+    CHIP_ERROR err                                 = CHIP_NO_ERROR;
+    chip::Transport::PeerMessageCounter * counter  = nullptr;
+    chip::Transport::PeerMessageCounter * counter2 = nullptr;
+    chip::Transport::GroupPeerTable mGroupPeerMsgCounter;
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(fabricIndex, peerNodeId, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter != nullptr);
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter2);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter2 != nullptr);
+    NL_TEST_ASSERT(inSuite, counter2 != counter);
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(fabricIndex, peerNodeId, true, counter2);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter2 == counter);
+}
+
+void CounterTrustFirstTest(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err                                = CHIP_NO_ERROR;
+    chip::Transport::PeerMessageCounter * counter = nullptr;
+    chip::Transport::GroupPeerTable mGroupPeerMsgCounter;
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter != nullptr);
+
+    err = counter->VerifyOrTrustFirst(5656);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    counter->Commit(5656);
+
+    err = counter->VerifyOrTrustFirst(5756);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    counter->Commit(5756);
+
+    err = counter->VerifyOrTrustFirst(4756);
+    NL_TEST_ASSERT(inSuite, err != CHIP_NO_ERROR);
+
+    // test sequential reception
+    err = counter->VerifyOrTrustFirst(5757);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    counter->Commit(5757);
+
+    err = counter->VerifyOrTrustFirst(5758);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    counter->Commit(5758);
+
+    err = counter->VerifyOrTrustFirst(5756);
+    NL_TEST_ASSERT(inSuite, err != CHIP_NO_ERROR);
+
+    // Test Roll over
+    err = mGroupPeerMsgCounter.FindOrAddPeer(1, 99, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter != nullptr);
+
+    err = counter->VerifyOrTrustFirst(UINT32_MAX - 1);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    counter->Commit(UINT32_MAX - 1);
+
+    err = counter->VerifyOrTrustFirst(5);
```

Comment:
So this happens to work totally by accident because the window bit at UINT32_MAX-6 is not set, right?

Need to have more careful tests here around the overflow condition.

Reply:
Added more test for overflow check.